File: android/src/main/java/com/carlrosales/capacitor/subscriptions/SubscriptionsPlugin.java
package com.carlrosales.capacitor.subscriptions;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import org.json.JSONArray;

@CapacitorPlugin(name = "Subscriptions")
public class SubscriptionsPlugin extends Plugin {

    private Subscriptions implementation;
    
    @Override
    public void load() {
        implementation = new Subscriptions(getContext());
    }

    @PluginMethod
    public void getActiveSubscriptionsCount(PluginCall call) {
        try {
            int count = implementation.getActiveSubscriptionsCount();
            
            JSObject ret = new JSObject();
            ret.put("count", count);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to get subscription count: " + e.getMessage());
        }
    }

    @PluginMethod
    public void getSubscriptions(PluginCall call) {
        try {
            JSONArray subscriptions = implementation.getSubscriptions();
            
            JSObject ret = new JSObject();
            ret.put("subscriptions", subscriptions);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to get subscriptions: " + e.getMessage());
        }
    }

    @PluginMethod
    public void isSubscriptionActive(PluginCall call) {
        String productId = call.getString("productId");
        
        if (productId == null || productId.trim().isEmpty()) {
            call.reject("productId is required and cannot be empty");
            return;
        }
        
        try {
            boolean isActive = implementation.isSubscriptionActive(productId);
            
            JSObject ret = new JSObject();
            ret.put("isActive", isActive);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to check subscription status: " + e.getMessage());
        }
    }
}

File: ios/Sources/SubscriptionsPlugin/SubscriptionsPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@available(iOS 15.0, *)
@objc(SubscriptionsPlugin)
public class SubscriptionsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SubscriptionsPlugin"
    public let jsName = "Subscriptions"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getActiveSubscriptionsCount", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSubscriptions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isSubscriptionActive", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = Subscriptions()

    @objc func getActiveSubscriptionsCount(_ call: CAPPluginCall) {
        do {
            let count = implementation.getActiveSubscriptionsCount()
            call.resolve([
                "count": count
            ])
        } catch {
            call.reject("Failed to get subscription count: \(error.localizedDescription)")
        }
    }
    
    @objc func getSubscriptions(_ call: CAPPluginCall) {
        do {
            let subscriptions = implementation.getSubscriptions()
            call.resolve([
                "subscriptions": subscriptions
            ])
        } catch {
            call.reject("Failed to get subscriptions: \(error.localizedDescription)")
        }
    }
    
    @objc func isSubscriptionActive(_ call: CAPPluginCall) {
        guard let productId = call.getString("productId"), !productId.isEmpty else {
            call.reject("productId is required and cannot be empty")
            return
        }
        
        do {
            let isActive = implementation.isSubscriptionActive(productId: productId)
            call.resolve([
                "isActive": isActive
            ])
        } catch {
            call.reject("Failed to check subscription status: \(error.localizedDescription)")
        }
    }
}
